### PR TITLE
update protobuf to 4.23.4

### PR DESCRIPTION
**Pull request by M-Waszkiewicz-Anaconda**
protobuf >= 4.0.0, required for snowflake dbt 

[upstream](https://github.com/protocolbuffers/protobuf/tree/v23.4)
replaced abseil-cpp git submodule with conda package
replaced jsoncpp git submodule with conda package
replaced gtest git submodule with conda package
removed unnecessary patches
based on  [CF](https://github.com/conda-forge/libprotobuf-feedstock) with small changes/tweaks

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,113 +1,153 @@
-{% set version = "3.20.3" %}
+# keep this without major version to let the bot pick it up
+{% set version = "23.4" %}
+# protobuf doesn't add the major version in the tag, it's defined per language in
+# https://github.com/protocolbuffers/protobuf/blob/main/version.json
```
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml]: They do... See https://github.com/protocolbuffers/protobuf/releases/tag/v4.23.4.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,113 +1,153 @@
...
-        - pkg-config  # [not win]
-        - unzip  # [not win]
-        - make  # [not win]
-        - sed  # [osx and arm64]
+        - cmake
+        - ninja
+        - patch                # [unix]
+        - m2-patch             # [win]
       host:
+        - gtest 1.14.0
```
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml:51]: gtest has a run export, which means that libprotobuf will have a runtime requirement on gtest. We probably don't want that.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,113 +1,153 @@
...
-        # breaks backwards compatibility and new SONAME each minor release
-        # https://abi-laboratory.pro/tracker/timeline/protobuf/
-        # One exception was that 3.6.1 was incompatible with 3.6.0
-        - {{ pin_subpackage('libprotobuf', max_pin='x.x') }}
+        # protobuf now (intentionally) increments the SOVER with every patch release, which
+        # breaks tools like grpc_plugin_cpp if they get the wrong libprotoc at runtime, see
+        # https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/4075
+        - {{ pin_subpackage('libprotobuf', max_pin='x.x.x') }}
+      ignore_run_exports_from:
+        - jsoncpp
```
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml:39]: Why do we ignore jsoncpp's run_exports?

Comment (M-Waszkiewicz-Anaconda) [reply recipe/meta.yaml:39]: not used in runtime

File: recipe/meta.yaml (hunk 2 above)
Comment (M-Waszkiewicz-Anaconda) [reply recipe/meta.yaml:51]: yes, added it to ignore_run_exports

File: recipe/meta.yaml (hunk 3 above)
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:39]: Please add a comment in the recipe so that it's clearer why these are needed.

**Review by skupr-anaconda — APPROVED** [review]
LGTM

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,113 +1,158 @@
...
     build:
       missing_dso_whitelist:  # [s390x]
         - $RPATH/ld64.so.1    # [s390x]
       run_exports:
-        # breaks backwards compatibility and new SONAME each minor release
-        # https://abi-laboratory.pro/tracker/timeline/protobuf/
-        # One exception was that 3.6.1 was incompatible with 3.6.0
-        - {{ pin_subpackage('libprotobuf', max_pin='x.x') }}
+        # protobuf now (intentionally) increments the SOVER with every patch release, which
+        # breaks tools like grpc_plugin_cpp if they get the wrong libprotoc at runtime, see
```
Comment (skupr-anaconda) [inline recipe/meta.yaml:34]: Should we additionally test such (and other) downstream packages to verify that they can be built fine? 
My thoughts are directed by this major version bump...

File: recipe/meta.yaml [hunk 5]
```diff
@@ -1,113 +1,158 @@
...
+        - test -f ${PREFIX}/lib/{{ each_lib }}.a                       # [unix]
+        - if not exist %LIBRARY_LIB%\{{ each_lib }}-static.lib exit 1  # [win]
+        {% endfor %}
+
+        # more CMake integration
+        - cd cmake_test
+        - cmake -GNinja $CMAKE_ARGS .                 # [unix]
+        # windows cannot link with mix debug and release version, so set to Release
+        - cmake -GNinja -DCMAKE_BUILD_TYPE=Release .  # [win]
+        - cmake --build .
```
Comment (boldorider4) [inline recipe/meta.yaml:152]: What is being tested here exactly?

Comment (M-Waszkiewicz-Anaconda) [reply recipe/meta.yaml:152]: beside protoc executable and static/shared libraries package provides cmake definitions which could be reused while using the this package for building with upstreams, we testing here just that, see also cmake_tests/CMakeLists.txt

Comment (boldorider4) [reply recipe/meta.yaml:152]: can you test the validity of these CMake definitions?

File: recipe/meta.yaml (hunk 4 above)
Comment (M-Waszkiewicz-Anaconda) [reply recipe/meta.yaml:34]: yes, doing it now, will push on the same channel

File: recipe/cmake_test/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,26 @@
+project(cf_dummy LANGUAGES C CXX)
+cmake_minimum_required(VERSION 3.12)
+# We need to specify CONFIG to make sure that CMake
+# uses protobuf-config.cmake instead of relying on its
+# own FindProtobuf.cmake that does not know about
+# abseil transitive depedendency and about C++ version
+# required, see https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/4075#issuecomment-1569242816
+find_package(Protobuf CONFIG REQUIRED)
```
Comment (M-Waszkiewicz-Anaconda) [inline recipe/cmake_test/CMakeLists.txt:8]: @boldorider4 here we testing validity if the find_package works and it builds correctly in tests then it means is valid.

**Review by JeanChristopheMorinPerso — APPROVED** [review]

**Review by boldorider4 — APPROVED** [review]
LGTM

**Review by ryanskeith — APPROVED** [review]
These all look reasonable to me.

**Pull request closed without merging**